File: courses/codes/cclv1-jr-class/ReindeerLinkedListSolution/ReindeerSchool.py
from Reindeer import Reindeer

class ReindeerSchool:

    def __init__(self):
        self.leader = None
        self.count = 0
        print("Reindeer School created.")

    def enrol(self, r):
        if (self.leader is None):
            self.leader = r
        else:
            temp = self.leader
            while temp.next is not None:
                temp = temp.next
            temp.next = r
        self.count += 1
        self.print_enrol_status(r.name)

    def search(self, s):
        temp = self.leader
        while temp is not None:
            if temp.name == s:
                return True
            temp = temp.next
        return False

    def drop(self, r):
        if self.count == 0 or not self.search(r.name):
            return ""
        name = ""
        temp = self.leader
        if r == temp:
            if self.count == 1:
                self.leader = None
            else:
                self.leader = r.next
        else:
            while temp.next is not None:
                if temp.next == r:
                    temp.next = r.next
                    break

        self.count -= 1
        self.print_drop_status(r.name)
        return r.name


    # ===========================================
    # Mainly for checking status; DO NOT MODIFY!
    # ===========================================

    def print_enrol_status(self, r_name):
        print("Reindeer " + r_name + " has been enrolled in the Reindeer School.")

    def print_drop_status(self, r_name):
        print("Reindeer " + r_name + " has been dropped from the Reindeer School.")

    def print_search_status(self, r_name):
        if self.search(r_name):
            print("Reindeer " + r_name + " is enrolled in the Reindeer School.")
        else:
            print("Reindeer " + r_name + " is not enrolled in the Reindeer School.")

    def print_count_status(self):
        if self.count == 0:
            print("There are currently " + str(self.count) + " reindeers enrolled in the Reindeer School.")
        else:
            print("There are currently " + str(self.count) + " reindeers enrolled in the Reindeer School, led by " + self.leader.name + ".")

    def print_reindeers(self):
        S = "["
        temp = self.leader
        while (temp is not None):
            S += temp.name + " --> "
            temp = temp.next
        if self.count > 0:
            S = S[:-5]
        S += "]"
        print("Reindeers: " + S)

File: courses/codes/cclv1-jr-class/ReindeerLinkedListSolution/ReindeerSchool.java

public class ReindeerSchool {
	
	private Reindeer leader;
	private int count;
	
	public ReindeerSchool() {
		this.leader = null;
		this.count = 0;
		System.out.println("Reindeer School created.");
	}
	
	public void enrol(Reindeer r) {
		if (this.leader == null) {
			this.leader = r;
		} else {
			Reindeer temp = this.leader;
			while (temp.next != null) {
				temp = temp.next;
			}
			temp.next = r;
		}
		this.count++;
		this.printEnrolStatus(r.name);
	}
	
	
	public String drop(Reindeer r) {
		if (this.count == 0 || !this.search(r.name)) {
			return "";
		}
		String name = "";
		Reindeer temp = this.leader;
		if (r.equals(temp)) {
			if (this.count == 1) {
				this.leader = null;
			} else {
				this.leader = r.next;
			}
		} else {
			while (temp.next != null) {
				if (temp.next.equals(r)) {
					temp.next = r.next;
					break;
				}
			}
		}
		
		this.count--;
		this.printDropStatus(r.name);
		return r.name;
	}
	
	public boolean search(String s) {
		Reindeer temp = this.leader;
		while (temp != null) {
			if (temp.name.equals(s)) {
				return true;
			}
			temp = temp.next;
		}
		return false;
	}

	// ===========================================
	//	Mainly for checking status; DO NOT MODIFY!
	// ===========================================
	
	public void printReindeers() {
		String S = "[";
		Reindeer temp = this.leader;
		while (temp != null) {
			S += temp.name + " --> ";
			temp = temp.next;
		}
		if (this.count > 0)
			S = S.substring(0, S.length()-5);
		S += "]";
		System.out.println("Reindeers: " + S);
	}
	
	public void printCountStatus() {
		
		if (this.count == 0) {
			System.out.println("There are currently " + this.count + " reindeers enrolled in the Reindeer School.");
		} else {
			System.out.println("There are currently " + this.count + " reindeers enrolled in the Reindeer School, led by " + this.leader.name + ".");
		}
		
	}
	
	public void printEnrolStatus(String rName) {
		System.out.println("Reindeer " + rName + " has been enrolled in the Reindeer School.");	
	}
	
	public void printDropStatus(String rName) {
		System.out.println("Reindeer " + rName + " has been dropped from the Reindeer School.");	
	}
	
	
	
	public void printSearchStatus(String rName) {
		
		if (this.search(rName)) {
			System.out.println("Reindeer " + rName + " is enrolled in the Reindeer School.");
		} else {
			System.out.println("Reindeer " + rName + " is not enrolled in the Reindeer School.");
		}
		
			
	}


}

